File: test/asteroid_decoder_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:underdeck_app/features/tools/asteroid/domain/asteroid_models.dart';

/// R12 — table tests for [AsteroidDecoder.analyze] / [validationRules], the
/// pure decoding logic behind the asteroid tool. Tables are built inline so no
/// asset bundle is needed.

AsteroidEntry _e(String name,
        {double? multiplier, int? value, bool? pvp}) =>
    AsteroidEntry(
      name: name,
      emoji: '*',
      multiplier: multiplier,
      value: value,
      pvp: pvp,
    );

/// A tables fixture with just enough keys to exercise value/alert math.
final _tables = AsteroidTables(
  type: {'1': _e('Rock')},
  size: {'2': _e('Large', multiplier: 2.0), '5': _e('Huge', multiplier: 1.0)},
  structure: {'3': _e('Sparse'), '5': _e('Dense')},
  salvage: {'4': _e('Scrap')},
  law: {'6': _e('Patrolled'), '0': _e('Lawless', pvp: true)},
  resource: {
    '6': _e('Star-Tar', value: 5),
    '7': _e('Iron', value: 10),
    '8': _e('Gold', value: 20),
    '9': _e('Rare Gas', value: 30),
  },
);

void main() {
  group('AsteroidDecoder.analyze', () {
    test('decodes positions, resource value, and a rare-gas alert', () {
      final a = AsteroidDecoder.analyze('123456789', _tables);
      expect(a.typeKey, '1');
      expect(a.type.name, 'Rock');
      expect(a.sizeKey, '2');
      expect(a.wealth, 5);
      expect(a.lawKey, '6');
      expect(a.resources.map((r) => r.key), ['7', '8', '9']);
      // (10 + 20 + 30) * multiplier 2.0 * wealth 5 = 600.
      expect(a.resourceValue, 600.0);
      // structure '3' < 5 -> no infra alert; law '6' -> no combat; has '9'.
      expect(a.alerts, hasLength(1));
      expect(a.alerts.single.level, AsteroidAlertLevel.high);
    });

    test('emits infra + star-tar + combat alerts for a lawless dense field', () {
      // struct '5' (>=5), law '0' (lawless+pvp), resources 6/6/6 (star-tar).
      final a = AsteroidDecoder.analyze('155030666', _tables);
      expect(a.wealth, 3);
      // (5 + 5 + 5) * multiplier 1.0 * wealth 3 = 45.
      expect(a.resourceValue, 45.0);
      final levels = a.alerts.map((x) => x.level).toList();
      expect(levels, [
        AsteroidAlertLevel.info, // significant infrastructure
        AsteroidAlertLevel.critical, // star-tar (lawKey 0 + resource 6)
        AsteroidAlertLevel.warning, // combat-enabled zone (pvp)
      ]);
      final starTar = a.alerts[1];
      expect(starTar.message, contains('3-3')); // count-wealth
    });

    test('unknown keys fall back to AsteroidEntry.unknown', () {
      final a = AsteroidDecoder.analyze('900000000', _tables);
      expect(a.type.name, 'Unknown');
      expect(a.size.name, 'Unknown');
      expect(a.resourceValue, 0.0); // unknown resources contribute value 0
    });

    test('rejects a non-9-length id', () {
      expect(
        () => AsteroidDecoder.analyze('12345', _tables),
        throwsA(isA<AsteroidDecodeException>()),
      );
    });

    test('rejects a non-numeric id', () {
      expect(
        () => AsteroidDecoder.analyze('12345678x', _tables),
        throwsA(isA<AsteroidDecodeException>()),
      );
    });
  });

  group('AsteroidDecoder.validationRules / isFullyValid', () {
    Map<String, bool> byId(String raw) => {
          for (final r in AsteroidDecoder.validationRules(raw)) r.id: r.isSatisfied,
        };

    test('a well-formed id satisfies every rule', () {
      expect(AsteroidDecoder.isFullyValid('123456789'), isTrue);
      final rules = byId('123456789');
      expect(rules.values.every((v) => v), isTrue);
    });

    test('non-digits fail digits/length and position rules', () {
      final rules = byId('12abc6789');
      expect(rules['digits'], isFalse);
      expect(rules['length'], isFalse); // length gate requires allDigits
      expect(AsteroidDecoder.isFullyValid('12abc6789'), isFalse);
    });

    test('wrong length fails the length rule only when digits pass', () {
      final rules = byId('12345678'); // 8 digits
      expect(rules['digits'], isTrue);
      expect(rules['length'], isFalse);
    });

    test('position 1 must be 1 (asteroid marker)', () {
      expect(byId('023456789')['type'], isFalse);
      expect(byId('123456789')['type'], isTrue);
    });

    test('a zero in a resource slot fails the rss rule', () {
      expect(byId('123456709')['rss'], isFalse); // position 8 is 0
      expect(byId('123456789')['rss'], isTrue);
    });

    test('an empty id satisfies nothing', () {
      expect(byId('').values.every((v) => !v), isTrue);
    });
  });
}

File: webapp/src/features/tools/asteroid/asteroidDecoder.test.ts
// @vitest-environment node
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import { FormatException } from '../../../core/errors';
import {
  analyze,
  formatAmount,
  isValidId,
  validate,
  type AsteroidTables,
} from './asteroidDecoder';

const tables = JSON.parse(
  readFileSync(
    fileURLToPath(new URL('../../../../public/catalog/asteroid_tables.json', import.meta.url)),
    'utf-8',
  ),
) as AsteroidTables;

describe('validate', () => {
  it('flags empty input as failing every rule', () => {
    const rules = validate('');
    expect(rules.every((r) => !r.ok)).toBe(true);
  });

  it('accepts a fully valid id', () => {
    expect(isValidId('195016321')).toBe(true);
  });

  it('rejects wrong length', () => {
    const rules = validate('19501632');
    expect(rules.find((r) => r.id === 'length')?.ok).toBe(false);
    expect(rules.find((r) => r.id === 'digits')?.ok).toBe(true);
  });

  it('rejects non-digits', () => {
    const rules = validate('19501632a');
    expect(rules.find((r) => r.id === 'digits')?.ok).toBe(false);
  });

  it('requires position 1 to be 1', () => {
    expect(validate('295016321').find((r) => r.id === 'type')?.ok).toBe(false);
  });

  it('requires size/wealth/resources to be non-zero', () => {
    expect(validate('105016321').find((r) => r.id === 'size')?.ok).toBe(false);
    expect(validate('195006321').find((r) => r.id === 'wealth')?.ok).toBe(false);
    expect(validate('195016301').find((r) => r.id === 'rss')?.ok).toBe(false);
  });
});

describe('analyze', () => {
  it('throws typed errors with exact copy', () => {
    expect(() => analyze('12345', tables)).toThrow(FormatException);
    expect(() => analyze('12345', tables)).toThrow('Asteroid ID must be exactly 9 digits.');
    expect(() => analyze('12345678x', tables)).toThrow('Asteroid ID must contain digits only.');
  });

  it('decodes 195016321 with resourceValue = 20 and only the infrastructure alert', () => {
    const r = analyze('195016321', tables);
    expect(r.type.name).toBe('Asteroid');
    expect(r.size.name).toBe('Hypermassive');
    expect(r.size.multiplier).toBe(5.0);
    expect(r.structure.name).toBe('Shelter');
    expect(r.structure.risk).toBe(3);
    expect(r.salvage.name).toBe('No salvage');
    expect(r.law.name).toBe('Manned ship');
    expect(r.wealth).toBe(1);
    expect(r.resources.map((x) => x.name)).toEqual(['Helium', 'Oxygen', 'Hydrogen']);
    // (2 + 1 + 1) × 5.0 × 1 = 20
    expect(r.resourceValue).toBe(20);
    expect(r.resourceValueText).toBe('20');
    expect(r.alerts.map((a) => a.level)).toEqual(['info']);
  });

  it('appends alerts in the exact order: info, high, critical, warning', () => {
    // d = 1 9 5 0 5 0 6 6 9 : structure 5, wealth 5, law 0 (pvp), rss 6/6/9
    const r = analyze('195050669', tables);
    expect(r.alerts.map((a) => a.level)).toEqual(['info', 'high', 'critical', 'warning']);
    const critical = r.alerts.find((a) => a.level === 'critical');
    // two resource digits equal 6, wealth is 5
    expect(critical?.message).toBe('Star-Tar deposits detected! Estimated harvest rate: 2-5');
    // (3 + 3 + 5) × 5.0 × 5 = 275
    expect(r.resourceValue).toBe(275);
  });

  it('falls back to the Unknown entry on a lookup miss (no throw)', () => {
    const r = analyze('295016321', tables);
    expect(r.type.name).toBe('Unknown');
    expect(r.type.emoji).toBe('?');
  });

  it('only fires the combat alert for law digit 0', () => {
    // law digit 8 is named PvP but pvp:false - no combat alert
    const r = analyze('195018321', tables);
    expect(r.law.name).toBe('PvP');
    expect(r.alerts.some((a) => a.level === 'warning')).toBe(false);
  });
});

describe('formatAmount', () => {
  it('shows integers without decimals and non-integers with one', () => {
    expect(formatAmount(20)).toBe('20');
    expect(formatAmount(2.5)).toBe('2.5');
    expect(formatAmount(1.0)).toBe('1');
  });
});
